velocity_modelling/scripts/plot_velocity_model.py

```diff
@@ -0,0 +1,224 @@
+from enum import Enum
+from pathlib import Path
+from typing import Annotated, Optional
+
+import numpy as np
+import pandas as pd
+import typer
+import yaml
+
+from pygmt_helper import plotting
+from qcore import grid
+from velocity_modelling import bounding_box
+
+app = typer.Typer()
+
+
+class VelocityModelComponent(str, Enum):
+    p_wave = "p_wave"
+    s_wave = "s_wave"
+    density = "rho"
+
+
+@app.command()
+def plot_velocity_model(
+    output_plot_path: Annotated[
+        Path,
+        typer.Argument(
+            help="Path to output velocity model plot.", writable=True, dir_okay=False
+        ),
+    ],
+    vm_params_ffp: Annotated[
+        Optional[Path], typer.Option(help="Path to VM params YAML", dir_okay=False)
+    ] = None,
+    velocity_model_ffp: Annotated[
+        Optional[Path],
+        typer.Option(
+            help="Path to velocity model directory", file_okay=False, exists=True
+        ),
+    ] = None,
+    resolution: Annotated[
+        Optional[float], typer.Option(help="Resolution of the velocity model")
+    ] = None,
+    centre_lat: Annotated[
+        Optional[float], typer.Option(help="Centre latitude", min=-90, max=90)
+    ] = None,
+    centre_lon: Annotated[
+        Optional[float], typer.Option(help="Centre longitude", min=-180, max=180)
+    ] = None,
+    bearing: Annotated[
+        Optional[float], typer.Option(help="Model bearing", min=0, max=360)
+    ] = None,
+    extent_x: Annotated[
+        Optional[float], typer.Option(help="Extent in x-direction", min=0)
+    ] = None,
+    extent_y: Annotated[
+        Optional[float], typer.Option(help="Extent in y-direction", min=0)
+    ] = None,
+    title: Annotated[Optional[str], typer.Option(help="Figure title")] = None,
+    latitude_pad: Annotated[
+        float, typer.Option(help="Latitude padding for figure (in degrees)", min=0)
+    ] = 0,
+    longitude_pad: Annotated[
+        float, typer.Option(help="Longitude padding for figure (in degrees)", min=0)
+    ] = 0,
+    dpi: Annotated[
+        float, typer.Option(help="Plot output DPI (higher is better)")
+    ] = 300,
+    component: Annotated[
+        VelocityModelComponent,
+        typer.Option(help="Velocity model component to overlay."),
+    ] = VelocityModelComponent.p_wave,
+    slice: Annotated[
+        int, typer.Option(help="z-level slice of velocity model to plot.")
+    ] = 0,
+    nz: Annotated[
+        Optional[int], typer.Option(help="Number of z-slices in velocity model")
+    ] = None,
+    transparency: Annotated[
+        int, typer.Option(help="Velocity model overlay transparency, (0 = opaque)")
+    ] = 80,
```

Comment:
So this is not `Optional[int]` because the default is not None?
I feel it is a little confusing (it may be just me) to have Optional and typer.Option at the same time. Do we really need this?

Reply:
The difference between `Option` and `Optional` is thus:

`Optional` is a type that exists in the standard library. It does not, in spite of it's name, have anything to do with CLI options or function keyword arguments. Instead, `var: Optional[T]` simply says that `var`  may have type `T` but it may also be `None`. This is very useful for type-checking, because it allows the type-checker to verify that you handle the `None` case in your source code. For example:
```python
def get_parameter(parameter_name: str) -> Optional[float]:
    ... # Source the parameter somewhere. Parameter may not exist so we may return None.
area = get_parameter('length') * get_parameter('width')
```
will fail to type-check with `mypy` (and will highlight in my editor as an error) because `get_parameter` has return type `Optional[float]` and `*` cannot multiply `NoneType` together. That is, **it correctly informs me I forgot to handle the possibility that my function failed and returned `None`**. Note the above is identical to

```python
def get_parameter(parameter_name: str) -> float | None:
    ... # Source the parameter somewhere. Parameter may not exist so we may return None.
area = get_parameter('length') * get_parameter('width')
```
It is just encouraged to use the standard library built-ins for the same job. Moreover similar types with the name `Optional` exist in many other programming languages, such as [C++](https://en.cppreference.com/w/cpp/utility/optional), [Rust](https://doc.rust-lang.org/std/option/index.html),  [OCaml](https://ocaml.org/manual/5.2/api/Option.html), Scala, Swift, etc. So the name is what it is because many programmers from other type-safe languages would expect it to be named `Optional`. I personally would not have called it that if I was implementing the Python type-system.

In the context of the plot velocity model script, and why the `Optional` is used, I wanted a script that could handle velocity model parameters from multiple places (manually specified, vm params, realisation file, etc). By defaulting `vm_params_ffp` and friends to `None` I can perform checks to see which of the parameters the user defined and source the velocity model parameters appropriately. **Because these variables have `Optional` types, type-checking will confirm that I have handled all the edge-cases correctly**. In the case of the `transparency` variable, I can assign a sensible default value of 80 and then allow the user to change that. It should not have type `Optional[int]` because passing `None` would cause the code to fail type-checking and the script to crash if run with `transparency=None`.